tiledb/sm/misc/types.h

```diff
@@ -178,7 +178,9 @@ class Range {
   /** Returns a pointer to the end of the range. */
   const void* end() const {
     auto end_pos = var_size_ ? range_start_size_ : range_.size() / 2;
-    return range_.empty() ? nullptr : &range_[end_pos];
+    return range_.empty() ?
```

Comment:
This is called a lot in code that's very performance dependent. Could you do range_.data() + end_pos to remove the extra conditional?

Reply:
I think that would be incorrect/unsafe code, failing to make obvious when the circumstance occurs, returning an address of 'something' that is not logically there without it being obvious, as it will likely (usually) be a valid address to an invalid 'something'.  (It might occasionally be an invalid reference as with the base64 issue in aws sdk showing up on mac.)
The actual instruction sequence (vs2019 RelWithDebInfo) generated for 'my' approach is one instruction longer, tho it does contain an additional conditional jump, see comparison (v1 data() + end_pos, v2 'my' version.
![range end v1 vs v2](https://user-images.githubusercontent.com/78012908/156601575-4b9659c9-bcb8-4c56-874c-f7b5fb5b269a.png)

Reply:
I know this is just one instruction but still it's called during very performance sensitive code.

As far as safety goes:
In the fixed size case, this will return the middle point of the range which is double the fixed size, so the condition will always be true.

In the var size case, the user should always use this value with end_size, which will return 0 when this condition is false. So wether we return nullptr or a pointer to the end of the range, the user should not use it.

Reply:
David's right about the correctness issue. The real problem with performance is that we've got one class (`Range`) whose member functions need to handle both fixed and variable length data. If we take that as a given, we could get better performance for fixed types by splitting the function at the top into two blocks, one for fixed-length and one for variable-length. The fixed-length block wouldn't need all the comparisons that the variable-length block will.

Reply:
I'm pretty sure the difference is insignificant, but I will try to obtain some numbers.

Reply:
some numbers and code mod used to obtain them...
```
original...
reached 10000000 in 20.9347, ui 10000000
...
reached 10000000 in 18.9192, ui 10000000
reached 10000000 in 19.3124, ui 10000000
reached 10000000 in 18.7263, ui 10000000

-------------
my initial change to avoid debug library trap, contained additional branch...
(first trial not copied, took 25.9xxx secs)
..then 2nd trial...
reached 10000000 in 18.8648, ui 10000000
reached 10000000 in 18.8001, ui 10000000
reached 10000000 in 19.7121, ui 10000000

------------------
lr changes...
reached 10000000 in 24.8284, ui 10000000
reached 10000000 in 18.7946, ui 10000000
reached 10000000 in 19.7197, ui 10000000
reached 10000000 in 18.8824, ui 10000000

-----------------
code added around unit-cppapi-array.cc:1317
to drive the above measurements, this is the location
that was causing the vs2019 debug builds to croak
    {
      uint64_t ntries = 10000000;
      uint64_t ui;
      auto startt = std::chrono::high_resolution_clock::now();
      for (ui = 0; ui < ntries; ++ui) {
        std::array<std::string, 2> range = query_r.range(0, 0);
      }
      auto endt = std::chrono::high_resolution_clock::now();
      std::chrono::duration<double> elapsed = endt - startt;

      printf(
          "reached %" PRIu64 " in %g, ui %" PRIu64 "\n",
          ntries,
          elapsed.count(),
          ui);
      __debugbreak();
    }
    std::array<std::string, 2> range = query_r.range(0, 0);
```